File: plugins/zig/guest.zig
const String = extern struct {
    ptr: [*]u8,
    len: usize,
};

extern fn my_plugin_host_api_greet(name: *String, ret: *String) void;
extern fn plugin_world_string_set(ret: *String, s: [*:0]const u8) void;

export fn exports_plugin_world_run(ret: *String) void {
    var name: String = undefined;
    plugin_world_string_set(&name, "Zig");
    my_plugin_host_api_greet(&name, ret);
}

File: plugins/nim/guest.nim
type
  PluginWorldString {.importc: "plugin_world_string_t", header: "plugin_world.h", bycopy.} = object
    p: pointer
    len: csize_t

proc greet(name, ret: ptr PluginWorldString) {.
  importc: "my_plugin_host_api_greet", header: "plugin_world.h".}

proc stringSet(ret: ptr PluginWorldString; s: cstring) {.
  importc: "plugin_world_string_set", header: "plugin_world.h".}

proc exports_plugin_world_run(ret: ptr PluginWorldString) {.exportc.} =
  var name: PluginWorldString
  stringSet(addr name, "Nim")
  greet(addr name, ret)
